noxfile.py

```diff
@@ -173,13 +175,34 @@ def lint(session: nox.sessions.Session):
         A `nox.sessions.Session` object.
 
     """
+    import yaml
+
     # Pip install the session requirements.
     session.install("pre-commit")
-    # Execute the pre-commit linting tasks.
-    cmd = ["pre-commit", "run", "--all-files"]
-    hooks = ["black", "blacken-docs", "flake8", "isort"]
+
+    # Load the pre-commit configuration YAML file.
+    with open(".pre-commit-config.yaml", "r") as fi:
+        config = yaml.load(fi, Loader=yaml.FullLoader)
+
+    # Ensure black and isort only perform a status check
+    # with a custom pre-commit configuration.
+    hooks = []
+    for entry in config["repos"]:
+        hook = Path(urlsplit(unquote(entry["repo"])).path).name
+        hooks.append(hook)
+        if hook in ["black", "isort"]:
+            # Enforce a "--check" for the hook.
+            entry["hooks"][0]["args"].insert(0, "--check")
```

Comment:
Previously, when we just had Black, it worked perfectly well to allow a 'full' Black operation, since that raises errors anyway and the file modification was isolated to the Cirrus instance. I'd much prefer to go back to that if possible, since it doesn't require this extra complexity. Is there a problem with this now we're also using isort?

Reply:
For CI we definitely only want to do a `--check` for both `black` and `isort` across all files.

There is no change in behaviour here for CI for this PR. We want `pre-commit` to return a status code failure for both the `black` and `isort` hooks when a non-compliance is detected, not for them to fix the non-compliancies, otherwise failures will go unnoticed.

However, the `black` and `isort` `pre-commit` hooks will automagically fix non-compliances for the developer on `git commit`.

Reply:
>We want `pre-commit` to return a status code failure for both the `black` and `isort` hooks when a non-compliance is detected

That's also achievable without the `--check` option, in which case there's no need for the YAML editing. I appreciate that without `--check` it will also modify the files, but that's irrelevant since it's on a Cirrus machine.

Reply:
I believe a key distinction here is that **pre-commit** raises exit code `1` on modification, even if the tool (**Black** or **isort**) does not.

Reply:
Okay, so I've verified the exact behaviour here, and `pre-commit` without the `--check` argument to both `black` and `isort` fixes the files but **also** returns a non-zero return code for the `pre-commit run --all-files` use case.

So I'm happy that it's safe not to inject the `--check` argument but still get the same behaviour i.e., a non-compliance failure.

Also, I guess if the developer is using `nox` to lint the code, then they get the same behaviour if they'd installed `pre-commit` as part of their workflow... so that now aligns.

Parsing the `pre-commit` config file still has value as we can now add or remove hooks and the `lint` task behaves appropriately, rather then only running fixed hooks from a static list - which can go stale.